File: parser/javascript/src/lexer.ts
import { Token, TokenType, AdfError } from './types';

/**
 * Lexer for ADF format - tokenizes line by line
 */
export class Lexer {
  private inMultiline = false;
  private multilineQuoteCount = 0;

  /**
   * Tokenize ADF text into tokens
   */
  tokenize(text: string): Token[] {
    const tokens: Token[] = [];
    const lines = text.split('\n');

    for (let i = 0; i < lines.length; i++) {
      const token = this.tokenizeLine(lines[i], i + 1);
      if (token) {
        tokens.push(token);
      }
    }

    return tokens;
  }

  /**
   * Tokenize a single line
   */
  private tokenizeLine(line: string, lineNumber: number): Token | null {
    // Handle multiline continuation
    if (this.inMultiline) {
      return this.handleMultilineContinuation(line, lineNumber);
    }

    // Blank line
    if (!line.trim()) {
      return {
        type: TokenType.BlankLine,
        lineNumber,
        rawLine: line,
      };
    }

    // Try to parse header
    const headerToken = this.tryParseHeader(line, lineNumber);
    if (headerToken) {
      return headerToken;
    }

    // Key-value pair
    if (line.includes('=')) {
      return this.parseKeyValue(line, lineNumber);
    }

    // Scalar value
    return {
      type: TokenType.ScalarValue,
      lineNumber,
      rawLine: line,
      value: line.trim(),
    };
  }

  /**
   * Try to parse a header line
   */
  private tryParseHeader(line: string, lineNumber: number): Token | null {
    const stripped = line.trim();

    if (!stripped.endsWith(':')) {
      return null;
    }

    let pathPart = stripped.slice(0, -1).trim();
    const isAbsolute = pathPart.startsWith('#');

    if (isAbsolute) {
      pathPart = pathPart.slice(1).trim();
    }

    // Root section
    if (!pathPart && isAbsolute) {
      return {
        type: TokenType.AbsoluteHeader,
        lineNumber,
        rawLine: line,
        path: '',
        isAbsolute: true,
      };
    }

    if (!pathPart) {
      return null;
    }

    if (!this.isValidPath(pathPart)) {
      return null;
    }

    return {
      type: isAbsolute ? TokenType.AbsoluteHeader : TokenType.RelativeHeader,
      lineNumber,
      rawLine: line,
      path: pathPart,
      isAbsolute,
    };
  }

  /**
   * Parse a key=value line
   */
  private parseKeyValue(line: string, lineNumber: number): Token {
    const equalsPos = line.indexOf('=');
    const rawKey = line.slice(0, equalsPos).trim();
    const rawValue = line.slice(equalsPos + 1).trimStart();

    const key = rawKey;

    // Check for multiline value
    const quoteCount = this.countLeadingQuotes(rawValue);
    if (quoteCount > 0) {
      this.inMultiline = true;
      this.multilineQuoteCount = quoteCount;

      // Check if it ends on same line
      if (rawValue.length > quoteCount * 2 && this.endsWithQuotes(rawValue, quoteCount)) {
        // Single-line quoted value
        this.inMultiline = false;
        const value = this.extractQuotedValue(rawValue, quoteCount);
        const constraint = this.extractConstraintAfterQuotes(rawValue, quoteCount);

        return {
          type: TokenType.KeyValue,
          lineNumber,
          rawLine: line,
          key,
          value,
          constraint,
        };
      } else {
        // Multiline start
        const content = rawValue.length > quoteCount ? rawValue.slice(quoteCount) : '';

        return {
          type: TokenType.MultilineStart,
          lineNumber,
          rawLine: line,
          key,
          value: content,
          quoteCount,
        };
      }
    }

    // Simple value
    const { value, constraint } = this.parseValueAndConstraint(rawValue);

    return {
      type: TokenType.KeyValue,
      lineNumber,
      rawLine: line,
      key,
      value,
      constraint,
    };
  }

  /**
   * Handle multiline continuation
   */
  private handleMultilineContinuation(line: string, lineNumber: number): Token {
    if (this.endsWithQuotes(line, this.multilineQuoteCount)) {
      this.inMultiline = false;
      const content = line.length >= this.multilineQuoteCount
        ? line.slice(0, -this.multilineQuoteCount).trimEnd()
        : '';
      const constraint = this.extractConstraintAfterLine(line, this.multilineQuoteCount);

      return {
        type: TokenType.MultilineEnd,
        lineNumber,
        rawLine: line,
        value: content,
        constraint,
      };
    }

    return {
      type: TokenType.MultilineContent,
      lineNumber,
      rawLine: line,
      value: line,
    };
  }

  /**
   * Check if path is valid
   */
  private isValidPath(path: string): boolean {
    if (!path) return true;

    const parts = path.split('.');
    for (const part of parts) {
      // Quoted key
      if (part.startsWith('"') && part.endsWith('"')) {
        continue;
      }
      // Regular key pattern
      if (!/^[A-Za-z0-9_]+$/.test(part)) {
        return false;
      }
    }
    return true;
  }

  /**
   * Count leading quotes
   */
  private countLeadingQuotes(s: string): number {
    let count = 0;
    for (const char of s) {
      if (char === '"') {
        count++;
      } else {
        break;
      }
    }
    return count;
  }

  /**
   * Check if string ends with N quotes
   */
  private endsWithQuotes(s: string, count: number): boolean {
    if (s.length < count) return false;
    return s.slice(-count) === '"'.repeat(count);
  }

  /**
   * Extract value from within quotes
   */
  private extractQuotedValue(s: string, quoteCount: number): string {
    if (s.length < quoteCount * 2) return '';
    return s.slice(quoteCount, -quoteCount);
  }

  /**
   * Extract constraint after quotes
   */
  private extractConstraintAfterQuotes(s: string, quoteCount: number): string | undefined {
    if (s.length <= quoteCount * 2) return undefined;
    const after = s.slice(s.length - quoteCount);
    return this.parseConstraint(after);
  }

  /**
   * Extract constraint after closing quotes on a line
   */
  private extractConstraintAfterLine(line: string, quoteCount: number): string | undefined {
    if (line.length < quoteCount) return undefined;
    const after = line.slice(line.length - quoteCount);
    return this.parseConstraint(after);
  }

  /**
   * Parse value and constraint
   */
  private parseValueAndConstraint(s: string): { value: string; constraint?: string } {
    const constraint = this.parseConstraint(s);
    if (constraint) {
      const parenPos = s.lastIndexOf('(');
      if (parenPos !== -1) {
        const value = s.slice(0, parenPos).trimEnd();
        return { value, constraint };
      }
    }
    return { value: s.trim() };
  }

  /**
   * Parse constraint from string
   */
  private parseConstraint(s: string): string | undefined {
    const trimmed = s.trim();
    if (!trimmed) return undefined;

    const openParen = trimmed.lastIndexOf('(');
    const closeParen = trimmed.lastIndexOf(')');

    if (openParen === -1 || closeParen === -1 || closeParen < openParen) {
      return undefined;
    }

    const constraint = trimmed.slice(openParen + 1, closeParen).trim();
    return constraint || undefined;
  }
}

File: parser/rust/src/lexer.rs
use crate::error::Result;

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    BlankLine,
    AbsoluteHeader,
    RelativeHeader,
    KeyValue,
    ScalarValue,
    MultilineStart,
    MultilineContent,
    MultilineEnd,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub line_number: usize,
    pub raw_line: String,
    pub path: Option<String>,
    pub is_absolute: Option<bool>,
    pub key: Option<String>,
    pub value: Option<String>,
    pub constraint: Option<String>,
    pub quote_count: Option<usize>,
}

impl Token {
    fn new(token_type: TokenType, line_number: usize, raw_line: String) -> Self {
        Token {
            token_type,
            line_number,
            raw_line,
            path: None,
            is_absolute: None,
            key: None,
            value: None,
            constraint: None,
            quote_count: None,
        }
    }
}

pub struct Lexer {
    in_multiline: bool,
    multiline_quote_count: usize,
}

impl Lexer {
    pub fn new() -> Self {
        Lexer {
            in_multiline: false,
            multiline_quote_count: 0,
        }
    }

    pub fn tokenize(&mut self, text: &str) -> Result<Vec<Token>> {
        let mut tokens = Vec::new();

        for (i, line) in text.lines().enumerate() {
            let line_number = i + 1;
            if let Some(token) = self.tokenize_line(line, line_number)? {
                tokens.push(token);
            }
        }

        Ok(tokens)
    }

    fn tokenize_line(&mut self, line: &str, line_number: usize) -> Result<Option<Token>> {
        // Handle multiline continuation
        if self.in_multiline {
            return Ok(Some(self.handle_multiline_continuation(
                line,
                line_number,
            )));
        }

        // Blank line
        if line.trim().is_empty() {
            return Ok(Some(Token::new(
                TokenType::BlankLine,
                line_number,
                line.to_string(),
            )));
        }

        // Try to parse header
        if let Some(token) = self.try_parse_header(line, line_number)? {
            return Ok(Some(token));
        }

        // Key-value pair
        if line.contains('=') {
            return Ok(Some(self.parse_key_value(line, line_number)?));
        }

        // Scalar value
        Ok(Some(Token {
            token_type: TokenType::ScalarValue,
            line_number,
            raw_line: line.to_string(),
            value: Some(line.trim().to_string()),
            ..Token::new(TokenType::ScalarValue, line_number, line.to_string())
        }))
    }

    fn try_parse_header(&self, line: &str, line_number: usize) -> Result<Option<Token>> {
        let stripped = line.trim();

        if !stripped.ends_with(':') {
            return Ok(None);
        }

        let mut path_part = stripped[..stripped.len() - 1].trim().to_string();
        let is_absolute = path_part.starts_with('#');

        if is_absolute {
            path_part = path_part[1..].trim().to_string();
        }

        // Root section
        if path_part.is_empty() && is_absolute {
            return Ok(Some(Token {
                token_type: TokenType::AbsoluteHeader,
                line_number,
                raw_line: line.to_string(),
                path: Some(String::new()),
                is_absolute: Some(true),
                ..Token::new(TokenType::AbsoluteHeader, line_number, line.to_string())
            }));
        }

        if path_part.is_empty() {
            return Ok(None);
        }

        if !Self::is_valid_path(&path_part) {
            return Ok(None);
        }

        Ok(Some(Token {
            token_type: if is_absolute {
                TokenType::AbsoluteHeader
            } else {
                TokenType::RelativeHeader
            },
            line_number,
            raw_line: line.to_string(),
            path: Some(path_part),
            is_absolute: Some(is_absolute),
            ..Token::new(
                if is_absolute {
                    TokenType::AbsoluteHeader
                } else {
                    TokenType::RelativeHeader
                },
                line_number,
                line.to_string(),
            )
        }))
    }

    fn parse_key_value(&mut self, line: &str, line_number: usize) -> Result<Token> {
        let equals_pos = line.find('=').unwrap();
        let raw_key = line[..equals_pos].trim();
        let raw_value = &line[equals_pos + 1..].trim_start();

        let key = raw_key.to_string();

        // Check for multiline value
        let quote_count = Self::count_leading_quotes(raw_value);
        if quote_count > 0 {
            self.in_multiline = true;
            self.multiline_quote_count = quote_count;

            // Check if it ends on same line
            if raw_value.len() > quote_count * 2
                && Self::ends_with_quotes(raw_value, quote_count)
            {
                // Single-line quoted value
                self.in_multiline = false;
                let value = Self::extract_quoted_value(raw_value, quote_count);
                let constraint = Self::extract_constraint_after_quotes(raw_value, quote_count);

                return Ok(Token {
                    token_type: TokenType::KeyValue,
                    line_number,
                    raw_line: line.to_string(),
                    key: Some(key),
                    value: Some(value),
                    constraint,
                    ..Token::new(TokenType::KeyValue, line_number, line.to_string())
                });
            } else {
                // Multiline start
                let content = if raw_value.len() > quote_count {
                    raw_value[quote_count..].to_string()
                } else {
                    String::new()
                };

                return Ok(Token {
                    token_type: TokenType::MultilineStart,
                    line_number,
                    raw_line: line.to_string(),
                    key: Some(key),
                    value: Some(content),
                    quote_count: Some(quote_count),
                    ..Token::new(TokenType::MultilineStart, line_number, line.to_string())
                });
            }
        }

        // Simple value
        let (value, constraint) = Self::parse_value_and_constraint(raw_value);

        Ok(Token {
            token_type: TokenType::KeyValue,
            line_number,
            raw_line: line.to_string(),
            key: Some(key),
            value: Some(value),
            constraint,
            ..Token::new(TokenType::KeyValue, line_number, line.to_string())
        })
    }

    fn handle_multiline_continuation(&mut self, line: &str, line_number: usize) -> Token {
        if Self::ends_with_quotes(line, self.multiline_quote_count) {
            self.in_multiline = false;
            let content = if line.len() >= self.multiline_quote_count {
                line[..line.len() - self.multiline_quote_count]
                    .trim_end()
                    .to_string()
            } else {
                String::new()
            };
            let constraint = Self::extract_constraint_after_line(line, self.multiline_quote_count);

            Token {
                token_type: TokenType::MultilineEnd,
                line_number,
                raw_line: line.to_string(),
                value: Some(content),
                constraint,
                ..Token::new(TokenType::MultilineEnd, line_number, line.to_string())
            }
        } else {
            Token {
                token_type: TokenType::MultilineContent,
                line_number,
                raw_line: line.to_string(),
                value: Some(line.to_string()),
                ..Token::new(TokenType::MultilineContent, line_number, line.to_string())
            }
        }
    }

    fn is_valid_path(path: &str) -> bool {
        if path.is_empty() {
            return true;
        }

        for part in path.split('.') {
            if part.starts_with('"') && part.ends_with('"') {
                continue; // Quoted key
            }
            if !part
                .chars()
                .all(|c| c.is_alphanumeric() || c == '_')
            {
                return false;
            }
        }
        true
    }

    fn count_leading_quotes(s: &str) -> usize {
        s.chars().take_while(|&c| c == '"').count()
    }

    fn ends_with_quotes(s: &str, count: usize) -> bool {
        if s.len() < count {
            return false;
        }
        s.chars().rev().take(count).all(|c| c == '"')
    }

    fn extract_quoted_value(s: &str, quote_count: usize) -> String {
        if s.len() < quote_count * 2 {
            return String::new();
        }
        s[quote_count..s.len() - quote_count].to_string()
    }

    fn extract_constraint_after_quotes(s: &str, quote_count: usize) -> Option<String> {
        if s.len() <= quote_count * 2 {
            return None;
        }
        let after = &s[s.len() - quote_count..];
        Self::parse_constraint(after)
    }

    fn extract_constraint_after_line(line: &str, quote_count: usize) -> Option<String> {
        if line.len() < quote_count {
            return None;
        }
        let after = &line[line.len() - quote_count..];
        Self::parse_constraint(after)
    }

    fn parse_value_and_constraint(s: &str) -> (String, Option<String>) {
        let constraint = Self::parse_constraint(s);
        if let Some(_) = &constraint {
            if let Some(paren_pos) = s.rfind('(') {
                let value = s[..paren_pos].trim_end().to_string();
                return (value, constraint);
            }
        }
        (s.trim().to_string(), None)
    }

    fn parse_constraint(s: &str) -> Option<String> {
        let trimmed = s.trim();
        if trimmed.is_empty() {
            return None;
        }

        let open_paren = trimmed.rfind('(')?;
        let close_paren = trimmed.rfind(')')?;

        if close_paren < open_paren {
            return None;
        }

        let constraint = trimmed[open_paren + 1..close_paren].trim();
        if constraint.is_empty() {
            None
        } else {
            Some(constraint.to_string())
        }
    }
}

impl Default for Lexer {
    fn default() -> Self {
        Self::new()
    }
}
